### ISSUE TITLE
Redmine Docker easy_gantt plugin installation problem

### ISSUE THREAD

**Issue by carrodher** [bitnami/bitnami-docker-redmine#96]
**Description**

Redmine Docker easy_gantt plugin installation problem

**Steps to reproduce the issue:**
Trying to install plugin (current version)

1. copyed unarchived plugin into /home/redmine/plugins
2. enterd redmine docker console by command docker exec -it redmine bash
3. changed dir to /bitnami/redmine/plugins/easy_gantt

**Describe the results you received:**

```
 bundle exec rake redmine:plugins:migrate RAILS_ENV=production

Could not find gem 'redmine_extensions' in any of the gem sources listed in your Gemfile.
Run `bundle install` to install missing gems.

root@213b5cbf2641:/bitnami/redmine/plugins/easy_gantt# bundle install
The latest bundler is 1.16.0.pre.3, but you are currently running 1.15.3.
To update, run `gem install bundler --pre`
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all non-root users on this machine.
Fetching gem metadata from*****ubygems.org/...........
Fetching version metadata from *******rubygems.org/..
Fetching dependency metadata from*****rubygems.org/.
Resolving dependencies...
Using rake 12.1.0
Using concurrent-ruby 1.0.5
Using i18n 0.8.6
Using minitest 5.10.3
Using thread_safe 0.3.6
Using builder 3.2.3
Using erubi 1.7.0
Using mini_portile2 2.3.0
Using crass 1.0.2
Using rack 2.0.3
Fetching nio4r 2.1.0
Installing nio4r 2.1.0 with native extensions
Using websocket-extensions 0.1.2
Using mime-types-data 3.2016.0521
Using arel 8.0.0
Using bundler 1.15.3
Using method_source 0.9.0
Using thor 0.20.0
Using tzinfo 1.2.3
Fetching nokogiri 1.8.1
Installing nokogiri 1.8.1 with native extensions
Using rack-test 0.7.0
Using sprockets 3.7.1
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /opt/bitnami/ruby/lib/ruby/gems/2.3.0/gems/nio4r-2.1.0/ext/nio4r
/opt/bitnami/ruby/bin/ruby -r ./siteconf20171009-999-1rt4g1r.rb extconf.rb
checking for unistd.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/opt/bitnami/ruby/bin/$(RUBY_BASE_NAME)
/opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:456:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:587:in `try_cpp'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:1091:in `block in have_header'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:942:in `block in checking_for'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:350:in `block (2 levels) in postpone'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:320:in `open'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:350:in `block in postpone'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:320:in `open'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:346:in `postpone'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:941:in `checking_for'
        from /opt/bitnami/ruby/lib/ruby/2.3.0/mkmf.rb:1090:in `have_header'
        from extconf.rb:5:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /opt/bitnami/ruby/lib/ruby/gems/2.3.0/extensions/x86_64-linux/2.3.0-static/nio4r-2.1.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /opt/bitnami/ruby/lib/ruby/gems/2.3.0/gems/nio4r-2.1.0 for inspection.
Results logged to /opt/bitnami/ruby/lib/ruby/gems/2.3.0/extensions/x86_64-linux/2.3.0-static/nio4r-2.1.0/gem_make.out

An error occurred while installing nio4r (2.1.0), and Bundler cannot continue.
Make sure that `gem install nio4r -v '2.1.0'` succeeds before bundling.

In Gemfile:
  redmine_extensions was resolved to 0.2.3, which depends on
    rails was resolved to 5.1.4, which depends on
      actioncable was resolved to 5.1.4, which depends on
        nio4r
root@213b5cbf2641:/bitnami/redmine/plugins/easy_gantt#
```

**Describe the results you expected:**

Plugin installed

**Comment by carrodher** [comment]
According to the[ easy_gantt docs](https://www.easyredmine.com/knowledge-base/gantt-knowledge/complete-manual?view=knowledge_detail&id=105#Commands%20to%20run%20after%20installation/update), we should run:
```
$ bundle install
$ bundle exec rake db:migrate RAILS_ENV=production
$ bundle exec rake redmine:plugins:migrate RAILS_ENV=production
```

But these commands are assuming that the base OS included some libraries and packages that we are not including in order to minimize the size of our images, so before running the previous commands, we should install these packages running
```
$ install_packages make ruby-dev gcc libmysqlclient-dev libmagickwand-dev imagemagick libpq-dev
```
Also, we can add these packages to the `Dockerfile` (5th line `RUN install_packages ...`).

TL;DR:
```
$ install_packages make ruby-dev gcc libmysqlclient-dev libmagickwand-dev imagemagick libpq-dev
$ cd /opt/bitnami/redmine/
$ bundle install
$ bundle exec rake db:migrate RAILS_ENV=production
$ bundle exec rake redmine:plugins:migrate RAILS_ENV=production
```
After that restart the container.

**Comment by NikolayUvarov** [comment]
Thank you for answer, but
after doing this 
`bundle exec rake db:migrate RAILS_ENV=production`
I have got:
```
You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

If this is a development machine, remove the /opt/bitnami/redmine/Gemfile freeze
by running `bundle install --no-deployment`.

The dependencies in your gemfile changed

You have added to the Gemfile:
* redmine_extensions

```
And installation fails.

**Comment by NikolayUvarov** [comment]
Bundle install also fails:

root@213b5cbf2641:/opt/bitnami/redmine# bundle install
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all non-root users on this
machine.
You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

If this is a development machine, remove the /opt/bitnami/redmine/Gemfile freeze
by running `bundle install --no-deployment`.

The dependencies in your gemfile changed

You have added to the Gemfile:
* redmine_extensions

**Comment by carrodher** [comment]
Do you try to run the command with the `--no-deployment` option as suggesting the above error?
```
bundle install --no-deployment
```

**Comment by NikolayUvarov** [comment]
Thanks, that helped, plugin installed. I had doubt if --no-deployment option is applicible to production.
But now plugin is displayed in top bar of redmine page.

Thank you very much for help!

**Comment by carrodher** [comment]
Hi, sorry for the delayed response. 
```
bundle install --deployment
```
will lock your `Gemfile.lock` and will not update it when you change your `Gemfile`, so never use deployment option in a development environment. 
```
bundle install --no-deployment
```
will disable bundle deployment mode.
